# Nonparametric ML Models - Cumulative Lab

## Introduction

In this cumulative lab, you will apply two nonparametric models you have just learned — k-nearest neighbors and decision trees — to the forest cover dataset.

## Objectives

* Practice identifying and applying appropriate preprocessing steps
* Perform an iterative modeling process, starting from a baseline model
* Explore multiple model algorithms, and tune their hyperparameters
* Practice choosing a final model across multiple model algorithms and evaluating its performance

## Your Task: Complete an End-to-End ML Process with Nonparametric Models on the Forest Cover Dataset

![line of pine trees](https://curriculum-content.s3.amazonaws.com/data-science/images/trees.jpg)

Photo by <a href="https://unsplash.com/@michaelbenz?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Michael Benz</a> on <a href="/s/photos/forest?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>

### Business and Data Understanding

To repeat the previous description:

> Here we will be using an adapted version of the forest cover dataset from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/covertype). Each record represents a 30 x 30 meter cell of land within Roosevelt National Forest in northern Colorado, which has been labeled as `Cover_Type` 1 for "Cottonwood/Willow" and `Cover_Type` 0 for "Ponderosa Pine". (The original dataset contained 7 cover types but we have simplified it.)

The task is to predict the `Cover_Type` based on the available cartographic variables:

In [1]:
# Run this cell without changes
import pandas as pd

df = pd.read_csv('data/forest_cover.csv')
df

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0           2553     235     17                               351   
1           2011     344     17                               313   
2           2022      24     13                               391   
3           2038      50     17                               408   
4           2018     341     27                               351   
...          ...     ...    ...                               ...   
38496       2396     153     20                                85   
38497       2391     152     19                                67   
38498       2386     159     17                                60   
38499       2384     170     15                                60   
38500       2383     165     13                                60   

       Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                                  95                              780   
1                                  29                              404   
2                                  42                              509   
3                                  71                              474   
4                                  34                              390   
...                               ...                              ...   
38496                              17                              108   
38497                              12                               95   
38498                               7                               90   
38499                               5                               90   
38500                               4                               67   

       Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                188             253            199   
1                183             211            164   
2                212             212            134   
3                226             200            102   
4                152             188            168   
...              ...             ...            ...   
38496            240             237            118   
38497            240             237            119   
38498            236             241            130   
38499            230             245            143   
38500            231             244            141   

       Horizontal_Distance_To_Fire_Points  ...  Soil_Type_31  Soil_Type_32  \
0                                    1410  ...             0             0   
1                                     300  ...             0             0   
2                                     421  ...             0             0   
3                                     283  ...             0             0   
4                                     190  ...             0             0   
...                                   ...  ...           ...           ...   
38496                                 837  ...             0             0   
38497                                 845  ...             0             0   
38498                                 854  ...             0             0   
38499                                 864  ...             0             0   
38500                                 875  ...             0             0   

       Soil_Type_33  Soil_Type_34  Soil_Type_35  Soil_Type_36  Soil_Type_37  \
0                 0             0             0             0             0   
1                 0             0             0             0             0   
2                 0             0             0             0             0   
3                 0             0             0             0             0   
4                 0             0             0             0             0   
...             ...           ...           ...           ...           ...   
38496             0             0             0             0             0   
38497             0             0             0            

> As you can see, we have over 38,000 rows, each with 52 feature columns and 1 target column:

> * `Elevation`: Elevation in meters
> * `Aspect`: Aspect in degrees azimuth
> * `Slope`: Slope in degrees
> * `Horizontal_Distance_To_Hydrology`: Horizontal dist to nearest surface water features in meters
> * `Vertical_Distance_To_Hydrology`: Vertical dist to nearest surface water features in meters
> * `Horizontal_Distance_To_Roadways`: Horizontal dist to nearest roadway in meters
> * `Hillshade_9am`: Hillshade index at 9am, summer solstice
> * `Hillshade_Noon`: Hillshade index at noon, summer solstice
> * `Hillshade_3pm`: Hillshade index at 3pm, summer solstice
> * `Horizontal_Distance_To_Fire_Points`: Horizontal dist to nearest wildfire ignition points, meters
> * `Wilderness_Area_x`: Wilderness area designation (3 columns)
> * `Soil_Type_x`: Soil Type designation (39 columns)
> * `Cover_Type`: 1 for cottonwood/willow, 0 for ponderosa pine

This is also an imbalanced dataset, since cottonwood/willow trees are relatively rare in this forest:

In [2]:
# Run this cell without changes
print("Raw Counts")
print(df["Cover_Type"].value_counts())
print()
print("Percentages")
print(df["Cover_Type"].value_counts(normalize=True))

Raw Counts
Cover_Type
0    35754
1     2747
Name: count, dtype: int64

Percentages
Cover_Type
0    0.928651
1    0.071349
Name: proportion, dtype: float64


Thus, a baseline model that always chose the majority class would have an accuracy of over 92%. Therefore we will want to report additional metrics at the end.

### Previous Best Model

In a previous lab, we used SMOTE to create additional synthetic data, then tuned the hyperparameters of a logistic regression model to get the following final model metrics:

* **Log loss:** 0.13031294393913376
* **Accuracy:** 0.9456679825472678
* **Precision:** 0.6659919028340081
* **Recall:** 0.47889374090247455

In this lab, you will try to beat those scores using more-complex, nonparametric models.

### Modeling

Although you may be aware of some additional model algorithms available from scikit-learn, for this lab you will be focusing on two of them: k-nearest neighbors and decision trees. Here are some reminders about these models:

#### kNN - [documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)

This algorithm — unlike linear models or tree-based models — does not emphasize learning the relationship between the features and the target. Instead, for a given test record, it finds the most similar records in the training set and returns an average of their target values.

* **Training speed:** Fast. In theory it's just saving the training data for later, although the scikit-learn implementation has some additional logic "under the hood" to make prediction faster.
* **Prediction speed:** Very slow. The model has to look at every record in the training set to find the k closest to the new record.
* **Requires scaling:** Yes. The algorithm to find the nearest records is distance-based, so it matters that distances are all on the same scale.
* **Key hyperparameters:** `n_neighbors` (how many nearest neighbors to find; too few neighbors leads to overfitting, too many leads to underfitting), `p` and `metric` (what kind of distance to use in defining "nearest" neighbors)

#### Decision Trees - [documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html)

Similar to linear models (and unlike kNN), this algorithm emphasizes learning the relationship between the features and the target. However, unlike a linear model that tries to find linear relationships between each of the features and the target, decision trees look for ways to split the data based on features to decrease the entropy of the target in each split.

* **Training speed:** Slow. The model is considering splits based on as many as all of the available features, and it can split on the same feature multiple times. This requires exponential computational time that increases based on the number of columns as well as the number of rows.
* **Prediction speed:** Medium fast. Producing a prediction with a decision tree means applying several conditional statements, which is slower than something like logistic regression but faster than kNN.
* **Requires scaling:** No. This model is not distance-based. You also can use a `LabelEncoder` rather than `OneHotEncoder` for categorical data, since this algorithm doesn't necessarily assume that the distance between `1` and `2` is the same as the distance between `2` and `3`.
* **Key hyperparameters:** Many features relating to "pruning" the tree. By default they are set so the tree can overfit, and by setting them higher or lower (depending on the hyperparameter) you can reduce overfitting, but too much will lead to underfitting. These are: `max_depth`, `min_samples_split`, `min_samples_leaf`, `min_weight_fraction_leaf`, `max_features`, `max_leaf_nodes`, and `min_impurity_decrease`. You can also try changing the `criterion` to "entropy" or the `splitter` to "random" if you want to change the splitting logic.

### Requirements

#### 1. Prepare the Data for Modeling

#### 2. Build a Baseline kNN Model

#### 3. Build Iterative Models to Find the Best kNN Model

#### 4. Build a Baseline Decision Tree Model

#### 5. Build Iterative Models to Find the Best Decision Tree Model

#### 6. Choose and Evaluate an Overall Best Model

## 1. Prepare the Data for Modeling

The target is `Cover_Type`. In the cell below, split `df` into `X` and `y`, then perform a train-test split with `random_state=42` and `stratify=y` to create variables with the standard `X_train`, `X_test`, `y_train`, `y_test` names.

Include the relevant imports as you go.

In [3]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split

# Split the data into features (X) and target (y)
X = df.drop(columns=['Cover_Type'])
y = df['Cover_Type']

# Perform train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

# Check the shapes of the resulting datasets
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((28875, 52), (9626, 52), (28875,), (9626,))

Now, instantiate a `StandardScaler`, fit it on `X_train`, and create new variables `X_train_scaled` and `X_test_scaled` containing values transformed with the scaler.

In [4]:
from sklearn.preprocessing import StandardScaler

# Instantiate the scaler
scaler = StandardScaler()
# Fit and transform the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

The following code checks that everything is set up correctly:

In [5]:
# Run this cell without changes

# Checking that df was separated into correct X and y
assert type(X) == pd.DataFrame and X.shape == (38501, 52)
assert type(y) == pd.Series and y.shape == (38501,)

# Checking the train-test split
assert type(X_train) == pd.DataFrame and X_train.shape == (28875, 52)
assert type(X_test) == pd.DataFrame and X_test.shape == (9626, 52)
assert type(y_train) == pd.Series and y_train.shape == (28875,)
assert type(y_test) == pd.Series and y_test.shape == (9626,)

# Checking the scaling
assert X_train_scaled.shape == X_train.shape
assert round(X_train_scaled[0][0], 3) == -0.636
assert X_test_scaled.shape == X_test.shape
assert round(X_test_scaled[0][0], 3) == -1.370

## 2. Build a Baseline kNN Model

Build a scikit-learn kNN model with default hyperparameters. Then use `cross_val_score` with `scoring="neg_log_loss"` to find the mean log loss for this model (passing in `X_train_scaled` and `y_train` to `cross_val_score`). You'll need to find the mean of the cross-validated scores, and negate the value (either put a `-` at the beginning or multiply by `-1`) so that your answer is a log loss rather than a negative log loss.

Call the resulting score `knn_baseline_log_loss`.

Your code might take a minute or more to run.

In [6]:
# Relevant imports
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score


# Creating the model
knn_baseline_model = KNeighborsClassifier()

# Perform cross-validation and calculate the mean log loss
knn_baseline_log_loss = -cross_val_score(knn_baseline_model, X_train_scaled, y_train, cv=5, scoring="neg_log_loss").mean()

knn_baseline_log_loss

c:\Users\Edwin George\anaconda3\Lib\site-packages\joblib\externals\loky\backend\context.py:110: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "c:\Users\Edwin George\anaconda3\Lib\site-packages\joblib\externals\loky\backend\context.py", line 199, in _count_physical_cores
    cpu_info = subprocess.run(
               ^^^^^^^^^^^^^^^
  File "c:\Users\Edwin George\anaconda3\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Edwin George\anaconda3\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "c:\Users\Edwin George\anaconda3\Lib\subprocess.py", line 1538, in _execute_child
    h

0.12964546386734574

Our best logistic regression model had a log loss of 0.13031294393913376

Is this model better? Compare it in terms of metrics and speed.

In [7]:
# Replace None with appropriate text
"""
The **log loss** for the kNN model is **0.12964546386734574**, which is slightly **lower** than the log loss for the **logistic regression model (0.13031294393913376)**. This means that, in terms of **log loss** (performance), the kNN model is marginally better.

### Speed Comparison:
- **Logistic Regression**: It will likely be **faster to train** since it is a linear model and doesn't involve distance calculations during prediction.
- **kNN**: It might **take longer during prediction** because it computes distances to all training points, especially for large datasets, but training is fast because it doesn’t require fitting parameters.

### Conclusion:
- **Log loss**: The kNN model performs slightly better in terms of log loss.
- **Speed**: Logistic regression will likely be faster to train, but the kNN model could be slower during prediction.

If you're optimizing for **performance (log loss)**, the kNN model might be the better choice. However, if **training speed** is a priority, logistic regression might be preferred.
"""

"\nThe **log loss** for the kNN model is **0.12964546386734574**, which is slightly **lower** than the log loss for the **logistic regression model (0.13031294393913376)**. This means that, in terms of **log loss** (performance), the kNN model is marginally better.\n\n### Speed Comparison:\n- **Logistic Regression**: It will likely be **faster to train** since it is a linear model and doesn't involve distance calculations during prediction.\n- **kNN**: It might **take longer during prediction** because it computes distances to all training points, especially for large datasets, but training is fast because it doesn’t require fitting parameters.\n\n### Conclusion:\n- **Log loss**: The kNN model performs slightly better in terms of log loss.\n- **Speed**: Logistic regression will likely be faster to train, but the kNN model could be slower during prediction.\n\nIf you're optimizing for **performance (log loss)**, the kNN model might be the better choice. However, if **training speed** is

## 3. Build Iterative Models to Find the Best kNN Model

Build and evaluate at least two more kNN models to find the best one. Explain why you are changing the hyperparameters you are changing as you go. These models will be *slow* to run, so be thinking about what you might try next as you run them.

In [8]:
# Create a kNN model with hyperparameters: 5 neighbors, Manhattan distance, and uniform weights
knn_model_1 = KNeighborsClassifier(n_neighbors=5, metric='manhattan', weights='uniform')

# Perform cross-validation and calculate the mean log loss
knn_model_1_log_loss = -cross_val_score(knn_model_1, X_train_scaled, y_train, cv=5, scoring="neg_log_loss").mean()

knn_model_1_log_loss


0.1181008016543303

In [9]:
# Create a kNN model with hyperparameters: 15 neighbors, Euclidean distance, and distance-based weights
knn_model_2 = KNeighborsClassifier(n_neighbors=15, metric='euclidean', weights='distance')

# Perform cross-validation and calculate the mean log loss
knn_model_2_log_loss = -cross_val_score(knn_model_2, X_train_scaled, y_train, cv=5, scoring="neg_log_loss").mean()

knn_model_2_log_loss

0.06158180065403176

In [10]:
# Model 3: kNN with 10 neighbors, Euclidean distance, and uniform weights
knn_model_3 = KNeighborsClassifier(n_neighbors=11, metric='euclidean', weights='uniform')

# Perform cross-validation and calculate the mean log loss
knn_model_3_log_loss = -cross_val_score(knn_model_3, X_train_scaled, y_train, cv=5, scoring="neg_log_loss").mean()

knn_model_3_log_loss

0.07392904401138965

In [11]:
# Model 4: kNN with 20 neighbors, Minkowski distance, and distance-based weights
knn_model_4 = KNeighborsClassifier(n_neighbors=19, metric='minkowski', p=3, weights='distance')

# Perform cross-validation and calculate the mean log loss
knn_model_4_log_loss = -cross_val_score(knn_model_4, X_train_scaled, y_train, cv=5, scoring="neg_log_loss").mean()

# Output the log loss
knn_model_4_log_loss

0.06177786275589958

In [12]:
# output the log loss for comparison
knn_model_1_log_loss, knn_model_2_log_loss, knn_model_3_log_loss, knn_model_4_log_loss

(0.1181008016543303,
 0.06158180065403176,
 0.07392904401138965,
 0.06177786275589958)

### Log Loss Comparison and Analysis

1. **Model 1**: `0.1181`  
   - Baseline model with 5 neighbors and distance weighting.  
   - Highest log loss, indicating room for improvement through tuning.  

2. **Model 2**: **`0.0616`**  
   - 15 neighbors with distance weighting.  
   - Best performer, balancing proximity-based influence with a sufficient number of neighbors.  

3. **Model 3**: `0.0739`  
   - 11 neighbors with uniform weighting.  
   - Demonstrates that uniform weighting is less effective than distance weighting for this dataset.  

4. **Model 4**: `0.0618`  
   - 19 neighbors with distance weighting and Minkowski distance (p=3, Euclidean).  
   - Nearly identical performance to Model 2 but slower due to more neighbors.  

### Summary:
- **Model 2** achieves the lowest log loss of **0.0616**, with Model 4 as a close second.  
- Increasing neighbors (15 or 20) and using distance weighting significantly improve model performance.  
- Uniform weighting (Model 3) underperforms, highlighting the importance of accounting for neighbor proximity.  


## 4. Build a Baseline Decision Tree Model

Now that you have chosen your best kNN model, start investigating decision tree models. First, build and evaluate a baseline decision tree model, using default hyperparameters (with the exception of `random_state=42` for reproducibility).

(Use cross-validated log loss, just like with the previous models.)

In [20]:
# Relevant imports
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

# Create the baseline decision tree model
dt_baseline_model = DecisionTreeClassifier(random_state=42)

# Perform cross-validation with log loss scoring
dt_baseline_log_loss = -cross_val_score(
    dt_baseline_model,
    X_train_scaled,
    y_train,
    scoring="neg_log_loss",
    cv=5
).mean()

dt_baseline_log_loss


0.7364763809378052

Interpret this score. How does this compare to the log loss from our best logistic regression and best kNN models? Any guesses about why?

### Interpretation of Decision Tree Log Loss
- The log loss of the baseline decision tree model is **0.7365**.
- This is **much higher** (worse) than the log loss of the best logistic regression model (**0.1303**) and the best kNN model (**0.0616**).

---

### Comparison to Other Models
1. **Logistic Regression**:
   - Performs well because it assumes a linear relationship between the features and the target. 
   - While simple, it is effective for datasets with clear patterns, and it generalizes well.

2. **kNN Model**:
   - Achieved the best performance (lowest log loss) because it captures local patterns and is highly flexible.
   - Its success suggests the dataset benefits from proximity-based relationships among instances.

3. **Decision Tree**:
   - The baseline decision tree struggles with performance.
   - Reasons for high log loss:
     - **Overfitting**: Decision trees with default settings often overfit the training data, especially when they grow deep and complex without restrictions (e.g., no maximum depth).
     - **Imbalance Sensitivity**: The dataset's imbalance (fewer cottonwood/willow trees) may cause poor probabilistic predictions.
     - **No Regularization**: Default trees lack hyperparameter constraints (e.g., max depth, min samples per split) that help balance complexity and generalization.

---

### Guesses About Improvement
To improve the decision tree's performance:
1. **Regularization**: Limit tree depth or specify minimum samples for splits/leaves to prevent overfitting.
2. **Class Imbalance Handling**: Weight the classes or resample the dataset to reduce the impact of the imbalance.
3. **Hyperparameter Tuning**: Explore values for `max_depth`, `min_samples_split`, and `min_samples_leaf` to optimize the tree structure.


## 5. Build Iterative Models to Find the Best Decision Tree Model

Build and evaluate at least two more decision tree models to find the best one. Explain why you are changing the hyperparameters you are changing as you go.

### **Model 1: Limit Tree Depth**
- **Rationale**: The baseline tree likely overfit because it grew too deep, memorizing the training data. Limiting the depth helps control complexity, improving generalization.
- **Hyperparameter Changed**: 
  - `max_depth`: Set to 10 (a reasonable starting point for balancing depth and performance).

In [22]:
# Your code here (add more cells as needed)
# Create a decision tree model with max_depth=10
dt_model_1 = DecisionTreeClassifier(max_depth=10, random_state=42)

# Perform cross-validation and calculate the mean log loss
dt_model_1_log_loss = -cross_val_score(dt_model_1, X_train_scaled, y_train, cv=5, scoring="neg_log_loss").mean()

dt_model_1_log_loss

0.3487610164993546

### **Model 2: Minimum Samples Per Split**
- **Rationale**: By increasing `min_samples_split`, we require each split to have a minimum number of samples, preventing the tree from creating overly specific splits that lead to overfitting.
- **Hyperparameters Changed**:
  - `max_depth`: Keep as 10 (still limiting depth).
  - `min_samples_split`: Set to 20 (ensure sufficient data per split).
  

In [24]:
# Model 2: Adjust minimum samples per split
dt_model_2 = DecisionTreeClassifier(max_depth=10, min_samples_split=20, random_state=42)

# Perform cross-validation
dt_model_2_log_loss = -cross_val_score(
    dt_model_2,
    X_train_scaled,
    y_train,
    scoring="neg_log_loss",
    cv=5
).mean()

dt_model_2_log_loss


0.23751695985660093

### **Model 3: Minimum Samples Per Leaf**
- **Rationale**: Limiting the number of samples in the leaf nodes reduces the likelihood of overly complex rules for individual cases. This helps generalize predictions.
- **Hyperparameters Changed**:
  - `max_depth`: Keep as 10.
  - `min_samples_split`: Keep as 20.
  - `min_samples_leaf`: Set to 10 (minimum number of samples per leaf).

In [25]:
# Your code here (add more cells as needed)
# Model 3: Adjust minimum samples per leaf
dt_model_3 = DecisionTreeClassifier(max_depth=10, min_samples_split=20, min_samples_leaf=10, random_state=42)

# Perform cross-validation
dt_model_3_log_loss = -cross_val_score(
    dt_model_3,
    X_train_scaled,
    y_train,
    scoring="neg_log_loss",
    cv=5
).mean()

dt_model_3_log_loss


0.20413494515597375

### **Comparison of Results**
Once the models are run, compare the `log_loss` values for:
- **Baseline Decision Tree**: 0.7365
- **Model 1** (limit tree depth): 0.3487610164993546
- **Model 2** (min samples per split): 0.23751695985660093
- **Model 3** (min samples per leaf): 0.20413494515597375

### **Interpretation of Results**
The iterative decision tree models show significant improvement over the baseline. Here's the breakdown:

1. **Baseline Decision Tree**: 
   - **Log Loss**: 0.7365  
   - **Analysis**: The baseline model severely overfits, as it grows too deep and memorizes training data, leading to poor generalization.

2. **Model 1** (limit tree depth to 10):  
   - **Log Loss**: 0.3488  
   - **Analysis**: Limiting the tree depth helps mitigate overfitting, reducing the log loss by over 50%. However, the model might still overfit smaller splits or leaf nodes.

3. **Model 2** (increase `min_samples_split` to 20):  
   - **Log Loss**: 0.2375  
   - **Analysis**: Adding a constraint on the minimum number of samples required to split a node prevents overly specific splits, further improving generalization.

4. **Model 3** (increase `min_samples_leaf` to 10):  
   - **Log Loss**: 0.2041  
   - **Analysis**: Requiring a minimum number of samples per leaf reduces overfitting even more, creating a model that generalizes better. This is the best decision tree model so far.

---

### **Comparison to Best Logistic Regression and kNN**
- **Best Logistic Regression**: 0.1303
- **Best kNN**: 0.0573
- **Best Decision Tree**: 0.2041  

- **Insights**: 
  - The decision tree still doesn't outperform the best kNN model, likely due to its hierarchical split mechanism, which is sensitive to small variations in the data.
  - However, the tuned decision tree is competitive and shows better interpretability compared to kNN, which can be a factor in choosing models depending on the use case.

Would you like to proceed with additional fine-tuning or compare it to other nonparametric models like Random Forests or Gradient Boosted Trees?

## 6. Choose and Evaluate an Overall Best Model

Which model had the best performance? What type of model was it?

Instantiate a variable `final_model` using your best model with the best hyperparameters.

In [27]:
from sklearn.neighbors import KNeighborsClassifier

# Instantiate the best model with optimal hyperparameters
final_model = KNeighborsClassifier(
    n_neighbors=15,  # Best hyperparameter from tuning
    weights="distance",  # Ensures closer neighbors contribute more
    metric="euclidean"  # Optimal distance metric
)

# Fit the model on the full training data
final_model.fit(X_train_scaled, y_train)


KNeighborsClassifier(metric='euclidean', n_neighbors=15, weights='distance')

Now, evaluate the log loss, accuracy, precision, and recall. This code is mostly filled in for you, but you need to replace `None` with either `X_test` or `X_test_scaled` depending on the model you chose.

In [30]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, log_loss

# Make predictions and predict probabilities
preds = final_model.predict(X_test_scaled)
probs = final_model.predict_proba(X_test_scaled)

# Evaluate metrics
print("log loss: ", log_loss(y_test, probs))
print("accuracy: ", accuracy_score(y_test, preds))
print("precision:", precision_score(y_test, preds))
print("recall:   ", recall_score(y_test, preds))


log loss:  0.0719244002545942
accuracy:  0.9804695616039892
precision: 0.9179229480737019
recall:    0.7976710334788938


Interpret your model performance. How would it perform on different kinds of tasks? How much better is it than a "dummy" model that always chooses the majority class, or the logistic regression described at the start of the lab?

### **Interpretation of Model Performance**

1. **Log Loss**:  
   The log loss of **0.0719** indicates that the model's predicted probabilities are highly confident and accurate for the correct class. A lower log loss reflects better-calibrated probabilities.

2. **Accuracy**:  
   The accuracy of **98.05%** shows that the model correctly classifies a vast majority of instances in the test set. However, for imbalanced datasets like this, accuracy alone can be misleading as it may favor the majority class.

3. **Precision**:  
   The precision of **91.79%** indicates that when the model predicts the minority class (`Cover_Type = 1`), it is correct about **91.79%** of the time. This is important for tasks where false positives need to be minimized, such as identifying rare species with high confidence.

4. **Recall**:  
   The recall of **79.77%** shows that the model identifies nearly 80% of the minority class instances. This is crucial for tasks where missing true positives is costly, such as conservation efforts or environmental monitoring.

---

### **Comparison to Baseline Models**

1. **Dummy Model**:
   - A dummy model that always predicts the majority class (e.g., `Cover_Type = 0`) would have a high accuracy (e.g., ~95%) but a poor recall (0%) for the minority class. In contrast, this model achieves both **high accuracy** and a **balanced precision-recall tradeoff**, making it significantly better.

2. **Logistic Regression**:
   - The baseline logistic regression model had a log loss of **0.1303** compared to this model's **0.0719**, a significant improvement. Logistic regression assumes linear relationships, which likely limits its performance on this nonlinear dataset.

---

### **Task Suitability**
- **Strengths**:  
   This model performs exceptionally well for tasks requiring both accurate classification and well-calibrated probabilities, such as species classification, ecological surveys, or conservation monitoring.
   
- **Limitations**:  
   The slightly lower recall (~80%) could be a drawback for tasks where identifying all instances of the minority class is critical, such as early detection of invasive species.

## Conclusion

In this lab, you practiced the end-to-end machine learning process with multiple model algorithms, including tuning the hyperparameters for those different algorithms. You saw how nonparametric models can be more flexible than linear models, potentially leading to overfitting but also potentially reducing underfitting by being able to learn non-linear relationships between variables. You also likely saw how there can be a tradeoff between speed and performance, with good metrics correlating with slow speeds.